src/Control/Optics/Linear/Internal.hs

```diff
@@ -6,31 +6,442 @@
+-- >
+-- > pairTupleIso :: Iso' (a,b) (Pair a b)
+-- > pairTupleIso = iso toTuple fromTuple
+-- >
+-- > pair1 :: Pair Int Int
+-- > pair1 = Pair 2 3
+-- >
+-- > setPairViaTuple :: (a,b) -> Pair a b -> Pair a b
+-- > setPairViaTuple newVal pair = set pairTupleIso newVal pair
+-- >
+-- > -- *module> setPairViaTuple (5,6) pair1
+-- > -- Pair 5 6
+--
+-- == Appendix: How do the types make sense?
+--
+-- === What are the optic types and why?
+--
+-- All optics are functions, called __optic functions__ of the form
+--
+-- > forall arr. (a `arr` b) -> (s `arr` t)
+--
+-- where @arr@ is some specific kind of arrow. Note that is is a rank-2 type
+-- (which just means that there is a @forall@ nested inside a @forall@).
+--
+-- The type of each optic suffices to write functions that go from the optics
+-- in this module to the standard optic types in the
+-- [lens package](https://hackage.haskell.org/package/lens).
+--
+-- === How do these types imply the sub-typing diagram above?
+--
+-- Let's look at a simple example and then describe the core principle
+-- that explains the sub-typing diagram above.
+--
+-- The example: any @Iso a b s t@ is a @Lens a b s t@.
+-- The types are essentially these:
+--
+-- > type Iso a b s t  = forall p. Profunctor p      => a `p` b -> s `p` t
+-- > type Lens a b s t = forall p. (Strong (,) () p) => a `p` b -> s `p` t
+--
+-- By the definition of @Strong@ in "Data.Profunctor.Linear", 
+-- any @p@ with a @Strong m u p@ instance also has a @Profunctor p@ instance.
+--
+-- Hence, we can write
+--
+-- > convert :: Iso a b s t -> Lens a b s t
+-- > convert iso a_p_b = iso a_p_b
+--
+-- This works because these are rank-2 types. The @iso@ is an optic function
+-- that transforms any profunctor computation of type 
+-- @Profunctor p => a `p` b@.  The second argument to @convert@ is @a_p_b@
+-- which is a computation of type @Strong (,) () p => a `p` b@. However, since
+-- the @Strong@ constraint implies @Profunctor@ constraint, we can give @a_p_b@
+-- to @iso@. We could not write @convert@ with the types going the other way.
+--
+-- More generally, since isomorphisms are functions that work on all
+-- profunctors, they are also functions that work on all profunctors
+-- with additional constraints. Moreover, there are more optic functions that
+-- work on all profunctors with that @Strong@ constraint since the assumption
+-- of further constraints allows other optic functions to be written. This is
+-- analogus to how in abstract algebra, a theorem on groups is still a theorem
+-- about abelian groups and in fact, the theorems that hold on abelian groups
+-- are a superset of the theorems that hold on just groups.
+--
+-- __Please take care to understand the principle at play:__
+--
+-- /If @D@ is a stricter constraint than @C@ (i.e., any @D@ instance has a @C@/
+-- /instance), then the rank-2 type @forall x. C x => f@ is a sub-type of/
+-- /@forall x. D x => f@/
+--
+-- === What do these optic types /mean/?
```

Comment:
I found this section a bit out of place. I think it spends too much space to say "the types are complicated and very general", which is not a very useful insight IMO.